aggregator/elem_base.go

```diff
@@ -239,5 +260,89 @@ func (e gaugeElemBase) NewLockedAggregation(_ Options, aggOpts raggregation.Opti
+) error {
+	if !aggTypes.IsValidForGauge() {
+		return fmt.Errorf("invalid aggregation types %s for Gauge", aggTypes.String())
+	}
+	return nil
+}
+
+func (e *gaugeElemBase) Close() {}
+
+type parsedPipeline struct {
+	// Whether the source pipeline contains derivative transformations at its head.
+	HasDerivativeTransform bool
+
+	// Sub-pipline containing only transformation operations from the head
+	// of the source pipeline this parsed pipeline was derived from.
+	Transformations applied.Pipeline
+
+	// Whether the source pipeline contains a rollup operation that is either at the
+	// head of the source pipeline or immediately following the transformation operations
+	// at the head of the source pipeline if any.
+	HasRollup bool
+
+	// Rollup operation that is either at the head of the source pipeline or
+	// immediately following the transformation operations at the head of the
+	// source pipeline if applicable.
+	Rollup applied.Rollup
+
+	// The remainder of the source pipeline after stripping the transformation
+	// and rollup operations from the head of the source pipeline.
+	Remainder applied.Pipeline
+}
+
+// parsePipeline parses the given pipeline and returns an error if the pipeline is invalid.
+// A valid pipeline should take the form of one of the following:
+// * Empty pipeline with no operations.
+// * Pipeline that starts with a rollup operation.
+// * Pipeline that starts with a transformation operation and contains at least one
+//   rollup operation. Additionally, the transformation derivative order computed from
+//   the list of transformations must be no more than the maximum transformation derivative
+//   order that is supported.
+func newParsedPipeline(pipeline applied.Pipeline) (parsedPipeline, error) {
+	if pipeline.IsEmpty() {
+		return parsedPipeline{}, nil
+	}
+	var (
+		firstRollupOpIdx              = -1
+		transformationDerivativeOrder int
+		numSteps                      = pipeline.NumSteps()
+	)
+	for i := 0; i < numSteps; i++ {
+		pipelineOp := pipeline.At(i)
+		if pipelineOp.Type != op.TransformationType && pipelineOp.Type != op.RollupType {
+			return parsedPipeline{}, fmt.Errorf("pipeline %v step %d has invalid operation type %v", pipeline, i, pipelineOp.Type)
+		}
+		if pipelineOp.Type == op.RollupType {
+			if firstRollupOpIdx == -1 {
+				firstRollupOpIdx = i
+			}
+		} else if firstRollupOpIdx == -1 {
+			// We only care about the transformation operations at the head of the pipeline
+			// before the first rollup operation since those are going to be processed locally.
+			transformOp := pipelineOp.Transformation
+			// A binary transformation is a transformation that computes first-order derivatives.
+			if transformOp.Type.IsBinaryTransform() {
+				transformationDerivativeOrder++
+			}
+		}
+	}
+	if firstRollupOpIdx == -1 {
+		return parsedPipeline{}, fmt.Errorf("pipeline %v has no rollup operations", pipeline)
+	}
+	if transformationDerivativeOrder > maxSupportedTransformationDerivativeOrer {
```

Comment:
can you elaborate what does this mean? could you provide a pipeline example that will error out in this case?

Reply:
Talked offline, so pipelines that compute higher order derivatives require keeping more states including the raw values and lower order derivatives, and as such are not supported by the current implementation. An example is "aggregate Last | perSecond | perSecond". Will add comments to clarify.